Add Pack operation to LinalgExt dialect

chelini opened the pull request:
Start revisiting the `linalgx.relayout` operation in lights of the discussions we had. The semantics I would like to have is: `tensor.expand_shape + tensor.transpose`

TODO: 
1. Dropping the affine map and use attributes only?
2. Do we want to add also `tensor.pad` semantics?
3. Operation name: `pack` or `relayout`?

rengolin commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,50 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+  DeclareOpInterfaceMethods<MemoryEffectsOpInterface>, 
+  SingleBlockImplicitTerminator<"::mlir::iree_compiler::IREE::LinalgExt::YieldOp">,
+]>{
+  let summary = "relayout - pack: tensor expand + tensor transpose";
```
Update summary, relayout ref is unnecessary. Would be nice to have an example here.

rengolin commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,50 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+]>{
+  let summary = "relayout - pack: tensor expand + tensor transpose";
+  
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
```
Do we want to support multiple inputs/outputs? What's even the semantics in that case? Do all tensors need to be the same shape, or just compatible pair-wise?

chelini replied:
I would limit to 1 input / 1 output.

rengolin replied:
me too

rengolin commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,66 @@ LogicalResult TopkOp::getResultTilePosition(
+// PackOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult PackOp::verify() { return success(); }
```
TODO here? We should perhaps verify shapes and maps, at least for static, and maybe some properties for dynamic.

rengolin commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,50 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+  
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
+                       Variadic<AnyShaped>:$outputs,
+                       AffineMapAttr:$packMap);
```
What's the semantics of the `packMap`?

rengolin commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,66 @@ LogicalResult TopkOp::getResultTilePosition(
+                                                   ValueRange ivs) {
+  assert(ivs.size() == getOutputRank());
+  AffineMap packMap = getPackMap();
+  assert(packMap.getNumDims() == getOutputRank() &&
```
Shouldn't this be in verify?

rengolin commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,55 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+  
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
+                       Variadic<AnyShaped>:$outputs,
+                       AffineMapAttr:$packMap);
```
Ideas for replacing the pack map with attributes:
* Add a n-block argument, for example `{ 0, 32, 32 }`, for blocking `NHW` into `NH'W'hw` with `h = 32` and `H' = H/n` (same for W).
* Add a transpose argument, for example `[0, 2, 1]`, for transposing the last two dimensions.

I'm not sure how the blocking args would play with the transpose args (ie. do we transpose than block or block then transpose?). For example, is `#block {32, 16}, #transpose {1, 0}` block size 16 for dim 1 before or after the transpose?

MaheshRavishankar replied:
I'd like to keep the size of the inner block mixed static and dynamic, like how offsets, sizes and strides are specified in extract and insert slices. So the op could be something like 

```
%1 = linalg_ext.pack %0 inner_tile_dims [%block_0, %block_1] : tensor<?x?xf32> to tensor<?x?x?x?xf32>
```

I am assuming when you mean transpose you mean the transpose for the inner block dimensions. For that a list of integer attributes that specify the transpose would be sufficient.

MaheshRavishankar replied:
Also wondering if we could change the implementation to not use `AffineMapAttr` to start with. We can have helper functions that build the `AffineMap` if needed (like it is done for expand/collapse_shape ops. It does make the implementation of lowering to scalar simpler (as I see below), but I have found `AffineMap` to be too cumbersome to use.

chelini replied:
> 

To give you more context these are the blocking we would like to support: https://github.com/iree-org/iree/pull/10481#discussion_r976032524. Transposing the inner block dimensions seems sufficient.

chelini replied:
And yes we want to drop the affineMap!

hanhanW commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,55 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
 }
 
+// TODO: is pack the right name? After all we are re-arranging the tensor.
+def IREELinalgExt_PackOp : IREELinalgExt_Op<"pack",[
```
Do we consider adding pad semantics to the op and rename it to linalg_ext.pad_and_pack op?

Weigh in @MaheshRavishankar since he's been looking into it recently.

chelini replied:
Thanks @hanhanW for the feedbacks. I never thought about padding (Nicolas made me think about it). He suggested the following semantic: `tensor.pack == tensor.pad + tensor.expand_shape + tensor.transpose`. While for the expansion and the transpose it is clear what attributes we need for padding not so. Perhaps introducing a flag, similar to `nofold` in `PadOp` that tells the operation to pad along the blocked dimension if the dimension is not a multiple of the blocking factor? Or pass directly the padding to the operation like `PadOp` does?

nicolasvasilache replied:
I think modeling on the pad op makes sense.
The nofold attribute will be useful, I would call it copy and blend it in the semantics, which will be in line with a more general way forward: https://discourse.llvm.org/t/rfc-structured-codegen-beyond-rectangular-arrays/64707#proposal-bufferization-copy-view-and-the-need-for-more-expressive-types-13

The pack name + optional padding region seem good enough for me to convey padding.
Renaming the op will be easy if it ever gets confusing.

hanhanW commented on llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/convert_to_loops.mlir:
```diff
@@ -760,3 +760,61 @@ func.func @topk_memref_optional(%input_values: memref<2x10xf32>, %out_values: me
+// CHECK-DAG: %[[step:.*]] = arith.constant 1 : index
+// CHECK-DAG: %[[ubC:.*]] = arith.constant 8 : index
+// CHECK-DAG: %[[block:.*]] = arith.constant 32 : index
+// CHECK: scf.for %[[N:.*]] = %[[lb]] to %[[ubN]] step %[[step]] {
+// CHECK:   scf.for %[[C:.*]] = %[[lb]] to %[[ubC]] step %[[step]] {
+// CHECK:     scf.for %[[n:.*]] = %[[lb]] to %[[block]] step %[[step]] {
+// CHECK:       scf.for %[[c:.*]] = %[[lb]] to %[[block]] step %[[step]] {
+// CHECK:         %[[applyMapI:.*]] = affine.apply #[[MAP]](%[[N]], %[[n]])
+// CHECK:         %[[applyMapJ:.*]] = affine.apply #[[MAP]](%[[C]], %[[c]])
+// CHECK:         %[[scalar:.*]] = memref.load %arg0[%[[applyMapI]], %[[applyMapJ]]] : memref<128x256xf32>
+// CHECK:         memref.store %[[scalar]], %arg1[%[[N]], %[[C]], %[[n]], %[[c]]] : memref<4x8x32x32xf32>
+// CHECK:       }
+// CHECK:     }
+// CHECK:   }
+// CHECK: }
```
Thanks for working on it. I'm more confident that we can implement tiling interface for the op after seeing the IR!

chelini replied:
Yes, that's why I wanted to show it to you guys! I think this it the right way of implementing this operation.

MaheshRavishankar commented:
Hey @chelini thanks for starting on this. I didn't get around to review it today. I'll take a look tonight or tomorrow morning for sure.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,74 @@ LogicalResult TopkOp::getResultTilePosition(
+  if (getNumOutputs() != 1)
+    return op->emitOpError("expects only 1 output operand");
+  AffineMap packMap = getPackMap();
+  if (packMap.getNumDims() != getOutputRank())
```
Probably need to build on this a bit. I think the input is of rank r,  the output rank is 2*r. 
Also the number of tile values specified should be `r`.

MaheshRavishankar reviewed (changes requested):
Thanks @chelini . Have a few comments. I am happy to work off of this. Let me know.

chelini commented:
> Thanks @chelini . Have a few comments. I am happy to work off of this. Let me know.

Hi @MaheshRavishankar if you don't mind I will like to give it a try and really (really...really) appreciates your feedback/comments along the way. 

Before starting couple of comments/expectations :

1. Should we start the development into LinalgExt? I assume this op will end up in the tensor dialect and share more traits with tensor ops instead of Linalg. I am fine bootstrapping the development in LinalgExt.
2. Transposition: We can start by blocking the tensor, and then when we are happy with the operation, add transposition on top of it, as you suggested by having an array of integers.
3. What is your opinion on packing? Nicolas suggested the following semantic: `tensor.pack == tensor.pad + tensor.expand_shape + tensor.transpose`. Truth be told, I never thought about padding and always assumed nice users passing blocks that perfectly dived the iteration domain.
4. This brings us to my next question: how do we want to handle partial tiles? 
5. Should we both support tensor and memref or have two ops 1 in the memref dialect and 1 in the tensor dialect similar to `expand/collapse_shape`?

rengolin commented:
> Truth be told, I never thought about padding and always assumed nice users passing blocks that perfectly divide the iteration domain.
> 4. This brings me to my next question: how do we want to handle partial tiles?

Without padding, blocking and tiling of dynamic shapes gets harder to track. Our current _"policy"_ to barf on non-multiple shapes only goes so far, we need to be able to handle those gracefully.

For example, one policy could be to always pad to the nearest upper multiple, another to peel from the nearest lower multiple. Peeling may be a predicated load of the end, or it may be something more complicated, it will depend on the shapes, the machines running it, etc.

Because padding adds non-trivial costs to multi-dimensional shapes, this has to be a user choice (well, our users are compiler writers, so them). I'm not sure what options we should support initially, but something that any machine can do, however inefficiently, would be a good default.

MaheshRavishankar commented:
> > Thanks @chelini . Have a few comments. I am happy to work off of this. Let me know.
> 
> Hi @MaheshRavishankar if you don't mind I will like to give it a try and really (really...really) appreciates your feedback/comments along the way.
> 

Thats really great! I'd happy to collaborate on this, and shape it to work for IREE as well. Part of the exploration here for me, is what is an IREE-ism, and what are the spec of the core operation. Without going into all the gory details of IREE, the main requirement is that the packing operation has to support the inner tile dimensions being a dynamic value (not static, so not carried as an attribute). That is a key requirement for IREE for this to fit in with IREEs layering. So hoping for a design point that can work equally well when you have the static information. So thats why having `inner_tile_dimensions [<list of mixed static/dynamic>]` values similar to `offsets`, `sizes` and `strides` of `tensor.extract_slices/tensor.insert_slices` makes sense. You can have a canonicalization that given static values for the inner_tile_dimensions propagates that to the return type. So these are valid specifications

```
%result = tensor.pack_and_pad %source inner_tile_dims [%a, %b] outs(%init : tensor<?x?x?x?xf32>) : tensor<?x?xf32> to tensor<?x?x?x?xf32>
```
or
```
%result = tensor.pack_and_pad %source inner_tile_dims [8, 4] outs(%init : tensor<?x?x8x4xf32>) : tensor<?x?xf32> to tensor<?x?x8x4xf32>
```

and you can go from former to latter if `%a == 8` and `%b == 4`.


> Couple of comments/expectations :
> 
> 1. Should we start the development into LinalgExt? I assume this op will end up in the tensor dialect and share more traits with tensor ops instead of Linalg. I am fine bootstrapping the development in LinalgExt.

Lets do this in `LinalgExt`. Two advantages, (1) we have a single op for both buffer and tensor semantics to start with, just like how Linalg ops work, so bufferization becomes easy. I do suspect that we only need the `tensor` version of the operation eventually, but we can make that determination later.

> 2. Transposition: We can start by blocking the tensor, and then when we are happy with the operation, add transposition on top of it. I like your suggestion of passing an array of integers.

I think transposition makes sense. We can add another list of integers that specifies the transposition (and make it optional). Start with the trivial case where transposition is just identity, and then adapt to the different transpositions. Really the transposition becomes important only when lower to loops. For the tiling I think that part can be ignored.

> 3. What is your opinion on packing? Nicolas suggested the following semantic: `tensor.pack == tensor.pad + tensor.expand_shape + tensor.transpose`. Truth be told, I never thought about padding and always assumed nice users passing blocks that perfectly divide the iteration domain.

The operation I have in mind does packing and padding. So takes your tensor, and adds necessary padding to convert the data into blocks that perfectly divide. (So maybe worth adding a padding value to the operation as an single scalar operand). 
Just FYI, the `pad` + `expand_shape` + `transpose` and I have classified it as a fail. Its the "sink representation" cause it either really hard to do anything meaningful with that.

> 4. This brings me to my next question: how do we want to handle partial tiles? (i.e., assuming the following input `tensor<8x8xf32>` and the following blocking factors [6, 6], what should be the output tensor shape? Should we bail out?

Seeing it as pack and pad, you would end up with a tensor of shape `<2x2x6x6xf32>` and the padding is added as part of the operation. So you always end up with a blocked layout.

> 5. Should we both support tensor and memref or have two ops 1 in the memref dialect and 1 in the tensor dialect similar to `expand/collapse_shape`? (see comment above about current design).

(see above)

hanhanW commented:
> I do suspect that we only need the tensor version of the operation eventually, but we can make that determination later.

Yes, I think eventually we only need tensor version, but we can start with having both. The cost is very small if we start it with LinalgExt dialect. Eventually, we should be able to vectorize the op. Then we won't need the memref version.

> The operation I have in mind does packing and padding. So takes your tensor, and adds necessary padding to convert the data into blocks that perfectly divide. (So maybe worth adding a padding value to the operation as an single scalar operand).

+1. If we have `%result = tensor.pack_and_pad %source inner_tile_dims [%a, %b] outs(%init : tensor<?x?x?x?xf32>) : tensor<?x?xf32> to tensor<?x?x?x?xf32>` form, we can compute the shape sizes. We can infer the shape sizes if %a and %b are constant values.

> This brings me to my next question: how do we want to handle partial tiles? (i.e., assuming the following input tensor<8x8xf32> and the following blocking factors [6, 6], what should be the output tensor shape? Should we bail out?
> Seeing it as pack and pad, you would end up with a tensor of shape <2x2x6x6xf32> and the padding is added as part of the operation. So you always end up with a blocked layout.

In my mind, we will add either `affine.min` or `scf.if` when lowering it to loops. If the op carries `inner_tile_dims`, we're able to compute the bounds and shape sizes in this case.

MaheshRavishankar commented:
> > I do suspect that we only need the tensor version of the operation eventually, but we can make that determination later.
> 
> Yes, I think eventually we only need tensor version, but we can start with having both. The cost is very small if we start it with LinalgExt dialect. Eventually, we should be able to vectorize the op. Then we won't need the memref version.
> 
> > The operation I have in mind does packing and padding. So takes your tensor, and adds necessary padding to convert the data into blocks that perfectly divide. (So maybe worth adding a padding value to the operation as an single scalar operand).
> 
> +1. If we have `%result = tensor.pack_and_pad %source inner_tile_dims [%a, %b] outs(%init : tensor<?x?x?x?xf32>) : tensor<?x?xf32> to tensor<?x?x?x?xf32>` form, we can compute the shape sizes. We can infer the shape sizes if %a and %b are constant values.
> 
> > This brings me to my next question: how do we want to handle partial tiles? (i.e., assuming the following input tensor<8x8xf32> and the following blocking factors [6, 6], what should be the output tensor shape? Should we bail out?
> > Seeing it as pack and pad, you would end up with a tensor of shape <2x2x6x6xf32> and the padding is added as part of the operation. So you always end up with a blocked layout.
> 
> In my mind, we will add either `affine.min` or `scf.if` when lowering it to loops. If the op carries `inner_tile_dims`, we're able to compute the bounds and shape sizes in this case.

Thats what I had in mind as well!

nicolasvasilache commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,61 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
+                       Variadic<AnyShaped>:$outputs,
+                       Variadic<Index>:$inner_tiles,
+                       I64ArrayAttr:$static_inner_tiles);
```
I would spell out the permutation explicitly as an array, we will likely want to express various reorderings.

nicolasvasilache commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,61 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+  let assemblyFormat = [{
+    attr-dict
+    `ins` `(` $inputs `:` type($inputs) `)`
+    `outs` `(` $outputs `:` type($outputs) `)` 
```
Hmm do we need this to be DPS ?
I am thinking of pack/unpack as similar to insert/extract ops that would produce new tensor values.
No need to change the current design if things already connect well, I'm just trying to think how we want things to look like in the fullness of time.

chelini replied:
Agree with you. At the tensor level should be 1 input, 1 result, 0 output. But I would like the op to work across tensor and memref for now (i.e., to see the loop generated). We can change later.

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
+LogicalResult PackOp::verify() {
+  // input rank = r
+  // output rank = 2 * r (input rank + tile rank).
+  // inner tiles = r
```
wrench: do you accept 0 to explicitly skip a dimension ?

MaheshRavishankar replied:
Do we really want that? Having `0` this way means it is pervasive in the code to check for `0` and do something else. Would a list of dimensions that are data tiled work?

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
+  return mixedInnerTiles;
+}
+
+SmallVector<StringRef> PackOp::getLoopIteratorTypes() {
```
This looks more like a linalg op than a tensor op.
Despite it living in linalg_ext for now, I am thinking that this would live close to `tensor.pad`.

nicolasvasilache replied:
ah I missed that it was part of the tilinginterface, please move to a comment-delimited region that says tiling interface.

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
+// https://mlir.llvm.org/doxygen/TensorInferTypeOpInterfaceImpl_8cpp_source.html
+/// Helper function to convert a vector of `OpFoldResult`s into a vector of
+/// `Value`s.
+static SmallVector<Value> getAsValues(OpBuilder &b, Location loc,
```
This should live in `StaticValueUtils.h` upstream

chelini replied:
I will open an upstream PR soon.

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
+      }));
+}
+
+// TODO: what if we don't have a full tile?
```
Just always return `makeComposedAffineApply(ceil(...))`

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
+
+// TODO: what if we don't have a full tile?
+// TODO: dynamic support.
+static OpFoldResult getBlockedOutputDimFromInputShape(
```
I would just add ```
OpFoldResult PadOp::buildOutputDim(builder, dim) {
 // get other info from op directly
}
```

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
+}
+
+static SmallVector<OpFoldResult>
+getOutputShapeFromInputShape(OpBuilder &builder, Location loc, Value input,
```
Same:  ```
SmallVector<OpFoldResult> PadOp::buildOutputShape(builder, dim);
```

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
+      }));
+}
+
+SmallVector<Range> PackOp::getIterationDomain(OpBuilder &builder) {
```
Delimit what is TilingInterfaceImpl with comments?

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/ConvertToLoops.cpp:
```diff
@@ -82,12 +82,15 @@ struct TilingInterfaceLowerToLoopsPattern : public RewritePattern {
-      return rewriter.notifyMatchFailure(
-          tilableOp, "lower to loops needs to have tensor semantics");
-    }
+    // Tensor semantics?
```
historically lower to loops needs to have buffer semantics but we can lift if upstream supports

nicolasvasilache commented on llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/convert_to_loops.mlir:
```diff
@@ -760,3 +760,56 @@ func.func @topk_memref_optional(%input_values: memref<2x10xf32>, %out_values: me
+// -----
+
+func.func @NC_to_NCNC(%arg0: memref<128x256xf32>, %arg1: memref<4x8x32x32xf32>) -> memref<4x8x32x32xf32> {  
+  iree_linalg_ext.pack ins(%arg0: memref<128x256xf32>) outs(%arg1: memref<4x8x32x32xf32>) inner_tiles [32, 32]
```
I'd go for something similar to:
```
iree_linalg_ext.pack %arg0 inner_tiles [32, 32]:  tensor<128x256xf32> -> tensor<4x8x32x32xf32>
iree_linalg_ext.pack %arg0 inner_tiles [32, 32] into %out: (memref<128x256xf32>, memref<4x8x32x32xf32>)
```

or a better spelling .. tensor is not DPS, memref takes the separately allocated buffer

nicolasvasilache commented:
Sorry I am late to the party as I missed this PR:

> This brings me to my next question: how do we want to handle partial tiles? (i.e., assuming the following input 
> tensor<8x8xf32> and the following blocking factors [6, 6], what should be the output tensor shape? Should 
> we bail out?

It seems to me the minor sizes should always be what is specified by the op. Padding occurs to ensure this (either with an explicit value or leaves garbage depending on the use case). Without padding explicitly, you are almost always guaranteed to have cache line splitting.

> Should we both support tensor and memref or have two ops 1 in the memref dialect and 1 in the tensor 
> dialect similar to expand/collapse_shape? (see comment above about current design).

This is what I would go after: the tensor op is not DPS and the memref op takes a previously allocated buffer to fill.

> Just FYI, the pad + expand_shape + transpose and I have classified it as a fail. Its the "sink representation" 
> cause it either really hard to do anything meaningful with that.

Careful not to conflate the spec with the impl.
The impl is the reason we want a higher-level op (that can additionally support dynamic tiles).
The semantics does not change though: it can still be described as a notional `pad + expand_shape + transpose` and it may be beneficial to lower to that at some point after tiling and fusion.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +435,80 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+    SmallVector<OpFoldResult> getMixedInnerTiles();
+
+    // Infer and return the shape of the output.
+    SmallVector<OpFoldResult> buildOutputShape(OpBuilder &builder);
```
I would rather implement this under the `ReifyRankedShapedTypeOpInterface`. (https://github.com/llvm/llvm-project/blob/0ca5993741877ab7fd27a251cbc1895bd092d5ee/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L202) . It doesnt return an `OpFoldResult` unfortunately (that was my plan to start with but there was push back). But you can use the conversions from `Value` to `OpFoldResult` to get by.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1447,152 @@ LogicalResult TopkOp::getResultTilePosition(
+// PackOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult PackOp::verify() {
```
I am assuming this is going to be flushed out once we have the op nailed down?

chelini replied:
Yes, just added some basic checks now. But still not complete.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/roundtrip.mlir:
```diff
@@ -643,3 +643,30 @@ func.func @topk_tensor_optional(%input_values: tensor<20x10x8x4xf32>) -> (tensor
+// CHECK: func.func @relayout(
+func.func @relayout(%arg0: memref<3x3xf32>, %arg1: memref<3x3xf32>) {
+  // CHECK: iree_linalg_ext.pack
+  iree_linalg_ext.pack %arg0 inner_tiles [1, 1] interchange = [0, 1, 2, 3] into %arg1 : (memref<3x3xf32> memref<3x3xf32>)
```
I'd expect `interchange` to be size of `inner_tiles`. I dont see much value in having interchange for the outer dimensions.

Also given the `inner_tiles` is `[1, 1]` , i'd expect the output shape to be `memref<3x3x1x1xf32>` Its extra dimensions, but the op does not need to be too smart about it?

nicolasvasilache commented on llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/convert_to_loops.mlir:
```diff
@@ -790,7 +790,7 @@ func.func @NC_to_NCnc(%arg0: memref<128x256xf32>, %arg1: memref<4x8x32x32xf32>)
 
 func.func @KC_to_KCck(%arg0: memref<128x256xf32>, %arg1: memref<8x4x32x32xf32>) {
-  iree_linalg_ext.pack %arg0 inner_tiles [32, 32] interchange = [1, 0] into %arg1 : (memref<128x256xf32> memref<8x4x32x32xf32>)
+  iree_linalg_ext.pack %arg0 dims_pos = [0, 1] inner_tiles = [32, 32] interchange = [1, 0] into %arg1 : (memref<128x256xf32> memref<8x4x32x32xf32>)
```
If you are going for dims_pos-like, I think you can drop the interchange?

Making a more convoluted examples
```
iree_linalg_ext.pack %arg0 tiled_dims = [2, 0] inner_tiles = [2, 32] into %arg1 : (memref<128x256x2x1000f32> memref<8x256x1x1000x 2x32xf32>)
```

tiled_dims represent the dimensions that get tiled and innert_tiles the final sizes ?

chelini replied:
yes, correct. `dim_pos` is the index of the dimension to tile while `inner_tiles` is the tiling factor.

chelini replied:
you want to express `interchange` using `dim_pos`?

chelini commented:
@MaheshRavishankar, @rengolin, @hanhanW and @nicolasvasilache. We are ready for another round of reviews. I addressed comments that have emerged so far. Currently, the operation has semantics: `tensor.expand_shape` + `tensor.transpose`. I removed the optional `transpose` integer vector. It turned out that it was not necessary (thanks @nicolasvasilache). We use `dims_pos` to carries which dimensions to tile, and the order determines if we need to interchange or not. `inner_tiles` represents the tile factors, which materialize as upper bounds for the point loops. Last but not least now `buildOutputShape` is under `reifyResultShapes` as suggested by @MaheshRavishankar.

Please let me know if something is not clear and if you have additional feedbacks.

Next steps:
1. Implement `unpack`.
2. Provide transforms operations to use `pack` and `unpack`.
3. Add pad semantics to `pack` and `unpack`.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +433,98 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+]>{
+  let summary = "pack operation"; 
+  let description = [{
+    The pack operation takes a shaped type as input, tiles, and transposes it
```
Nit:  I think it would be easier to say something like
```
The pack operation converts an `input` into a tiled and packed layout. The dimensions to be tiled are obtained from `dims_pos` and the size of the tile is obtained from `inner_tiles`. The dimensions listed in `dims_pos` need not be contiguous in which case the tile will get transposed.
```

Having some IR examples for different cases would also be useful.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +433,98 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+    constraint in the future when we add pad semantics to the operation.
+  }];
+
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
```
Is there a reason to have `Variadic<AnyShaped>` instead of just a single `input`?

chelini replied:
You are right, it should be a single input and single output operation. I went for the variadic because `IREELinalgExt_Op` implements the `AttrSizedOperandSegments`. Other ops like `ReverseOp` do the same. I can definitely pull out `AttrSizedOperandSegments` from `IREELinalgExt_Op` and have a single input/output operand. I would like to do that in a follow-up PR.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +433,98 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+  let extraClassDeclaration = extraLinalgExtOpClassDeclaration # [{
+
+    // Return the output operand.
+    Value getOutput() {
```
If you have a single input/output, this and the next method get auto-generated.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:
```diff
@@ -435,6 +433,98 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+
+    // Return a mapping from positions `dims_pos` to their tile factors.
+    DenseMap<int64_t, OpFoldResult> getDimAndTileMapping();
+    DenseMap<int64_t, int64_t> getDimAndStaticTileMapping();
```
I am not sure we need the `getDimAndStaticTileMapping`. The `getDimAndTileMapping` has everything we need I think.

chelini replied:
removed.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+  for (auto i : llvm::seq<int64_t>(0, rank)) {
+    if (tileAndPosMapping.count(i)) {
+      if (inputType.isDynamicDim(i) ||
+          tileAndPosMapping[i] == ShapedType::kDynamicSize)
```
If you used `getDimAndTileMapping` you could change this

```
for (auto i : ) {
   if (tileAndPosMapping.count(i)) {
      Optional<int64_t> tileSize = getConstantIntValue(tileAndPosMapping.lookup(i));
      // Drop all `== ShapedType::kDynamicSize` everywhere.
   }
}
```

The `getConstantIntValue` is defined in `StaticValueUtils.h` in core

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+    return false;
+  };
+
+  return llvm::any_of(tiles, [&](OpFoldResult tile) { return isZero(tile); });
```
You can just replace `isZero` with `isConstantIntValue` from `StaticValueUtils.h`

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+  SmallVector<int64_t> dimsPos = extractFromI64ArrayAttr(getDimsPos());
+  // Blocking factors must be less or equal than the input rank, and must
+  // match the number of `dims_pos`.
+  if (numberOfBlockingFactors > getInputRank())
```
Nit : if body is multiple lines use `{` and `}` (that is LLVM style AFAIK). Here and everywhere else.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+
+  // The number of `dim_pos` that carries the index of the dimensions to block
+  // must be less or equal than the input rank.
+  if (dimsPos.size() > getInputRank())
```
Seems like a redundant check as compared to the above two `if` conditions.

chelini replied:
Agree. removed.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+
+// Implement the tiling interface. The number of loops equals
+// the rank of the output tensors. All the loops are parallel.
+SmallVector<StringRef> PackOp::getLoopIteratorTypes() {
```
I have to think about how this works out in practice. If we want this op to be "tilable", I expect the number of iterations to be same as the dimensionality of the input tensor rather than the output tensor.

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:
```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+}
+
+// Implements `getIterationDomain` from the tiling interface.
+LogicalResult PackOp::generateScalarImplementation(OpBuilder &builder,
```
I think if `getLoopIteratorType` returning just as many parallel loops as the input tensor type will make this change to have the `scf.for` for the tiled dimensions of the output being generated explicitly. This doesnt have to be part of this PR. It can be done after the fact (and I am happy to pick it up as well).
This way of doing it will make it compatible with implementing a tiling algorithm for this operation as well.

MaheshRavishankar reviewed (changes requested):
Great! Mostly looks ready to land. Have a few remaining nits. Have one major comment w.r.t iterator types and scalar implementation that can come after this PR.

hanhanW reviewed (commented):
Overall looks good! The main concern is covered by Mahesh.

I can help work on adding pad semantics to the ops after landing the PR.
